drivers/i3c/i3c_i2c_virtual_controller.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2020 Google LLC
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ * Copyright 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT virtual_i3c_i2c_controller
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(i3c_i2c_virtual_controller, CONFIG_I3C_LOG_LEVEL);
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i3c.h>
+
+struct i2c_virt_ctrl_data {
+	/* Pointer to the array of I2C devices */
+	struct i3c_i2c_device_desc *i2c_devs;
+
+	/* Number of I2C devices */
+	const size_t num_i2c_devs;
+};
+
+static struct i3c_i2c_device_desc *find_i2c_dev(const struct i2c_virt_ctrl_data *data,
+						uint8_t addr)
+{
+	int i;
+
+	for (i = 0; i < data->num_i2c_devs; i++) {
+		if (data->i2c_devs[i].addr == addr) {
+			return &data->i2c_devs[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int i2c_virt_ctrl_configure(const struct device *dev, uint32_t dev_config)
+{
+	return 0;
+}
+
+static int i2c_virt_ctrl_transfer(const struct device *dev, struct i2c_msg *msgs,
+				  uint8_t num_msgs, uint16_t addr)
+{
+	const struct i2c_virt_ctrl_data *data = dev->data;
+	struct i3c_i2c_device_desc *i2c_dev;
+
+	i2c_dev = find_i2c_dev(data, addr);
+
+	if (i2c_dev == NULL) {
+		LOG_ERR("Cannot find I2C device 0x%02x", addr);
+		return -ENODEV;
+	}
+
+	return i3c_i2c_transfer(i2c_dev, msgs, num_msgs);
+}
+
+/**
+ * Set up a new emulator and add it to the list
+ *
+ * @param dev I2C emulation controller device
+ */
+static int i2c_virt_ctrl_init(const struct device *dev)
+{
+	const struct i2c_virt_ctrl_data *data = dev->data;
+
+	int i;
+
+	for (i = 0; i < data->num_i2c_devs; i++) {
+		i3c_i2c_device_register(&data->i2c_devs[i]);
+	}
+
+	return 0;
+}
+
+static struct i2c_driver_api i2c_virt_ctrl_api = {
+	.configure = i2c_virt_ctrl_configure,
+	.transfer = i2c_virt_ctrl_transfer,
+};
+
+#define I2C_DEVICE_DESC(node_id)					\
+	{								\
+		.bus = DEVICE_DT_GET(DT_BUS(DT_PARENT(node_id))),	\
+		.addr = DT_PROP_BY_IDX(node_id, reg, 0),		\
+		.lvr = DT_PROP(DT_PARENT(node_id), lvr),		\
```

Comment:
does this really need to be a common property for all devices within the parent node? LVR should be with the device and not a common thing. It should be able to read it within the reg or as a new prop of the i2c-device

Reply:
The issue here is the usage of `DT_REG_ADDR()` to get the target address. If `reg` has 3 elements, this results in the address macro being a 96-bit integer, which cannot be used at all. To workaround that, `reg` remains a one element property, and a common LVR is introduced. If per device LVR is really important, the device driver will need to be modified to use the I3C API to define I2C devices.

Reply:
you could add `lvr` into i2c-device.yaml. It would just have to be `required: false` and `default: 0`. I don't see why a device driver would need to modified in that case.

Reply:
> The issue here is the usage of `DT_REG_ADDR()` to get the target address. If `reg` has 3 elements, this results in the address macro being a 96-bit integer, which cannot be used at all.

When I run `west build -p -b mimxrt685_evk_cm33 samples/sensor/lsm6dso_i3c/` and look at `devicetree_unfixed`. I see that i3c devices have the same issue.

```c
#define DT_N_S_soc_S_peripheral_50000000_S_i3c_36000_S_lsm6dso_6b_REG_IDX_0_VAL_ADDRESS 1973801618120312098827 /* 0x6b00000208006c100b */
```

Reply:
pinging on this^?

Reply:
I try to avoid adding anything to the I2C subsys to make I3C work, because those are not I2C features.

Reply:
> The issue here is the usage of DT_REG_ADDR() to get the target address. If reg has 3 elements, this results in the address macro being a 96-bit integer, which cannot be used at all. To workaround that, reg remains a one element property, and a common LVR is introduced.

I don't know what an LVR is, but have you seen how PCIe bus addresses work around this using ranges? See 1e8f0f3bd16758aa5a510b8711f919a2e0a1229f in particular. If `reg` is the more natural place for this information than a special property, you may be able to do a similar trick by matching against 'i3c' in the node bus.

Reply:
LVR (Legacy Virtual Register) is a 8-bit integer which describes the I2C device (e.g. mostly about how tolerant the device on SCL's speed).

Thanks for the link. I will look into it.